pkg/newstorage/newstorage.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package newstorage
+
+import "errors"
+
+// ErrStoreNotFound is returned when a store is not found.
+var ErrStoreNotFound = errors.New("store not found")
+
+// ErrDataNotFound is returned when data is not found.
+var ErrDataNotFound = errors.New("data not found")
+
+// StoreConfiguration represents the configuration of a store.
+type StoreConfiguration struct {
+	TagNames []string
+}
+
+// QueryOptions represents various options for Query calls in a store.
+type QueryOptions struct {
+	pageSize int
+}
+
+// QueryOption represents an option for a Query call in a store.
+type QueryOption func(opts *QueryOptions)
+
+// WithPageSize sets the maximum page size for data retrievals done within the Iterator returned by the Query call.
+// Paging is handled internally by the Iterator. Higher values may reduce CPU time and the number of database calls at
+// the expense of higher memory usage.
+func WithPageSize(size int) QueryOption {
+	return func(opts *QueryOptions) {
+		opts.pageSize = size
+	}
+}
+
+// Tag represents a Name + Value pair that can be associated with a k,v pair for querying later.
+type Tag struct {
+	Name  string
+	Value string
+}
+
+// Operation represents an operation to be performed in the Batch method.
+type Operation struct {
+	Key   string
+	Value []byte // A nil value will result in a delete operation
+	Tags  []Tag  // Optional
+}
+
+// Provider represents a storage provider.
+type Provider interface {
+	// OpenStore opens a store with the given name and returns a handle.
+	// If the store has never been opened before, then it is created.
+	OpenStore(name string) (Store, error)
+
+	// SetStoreConfig sets the configuration on a store.
+	// The store must be created prior to calling this method.
+	SetStoreConfig(name string, config StoreConfiguration) error
+
+	// GetStoreConfig gets the current store configuration.
+	// The store must be created prior to calling this method.
+	GetStoreConfig(name string) (StoreConfiguration, error)
+
+	// CloseStore closes the store with the given name.
+	// This does not delete any data in the store, and the store can be reopened again.
+	CloseStore(name string) error
+
+	// Close closes all stores created under this store provider.
+	// This does not delete any data in the stores, and the stores can be reopened again.
+	Close() error
+}
+
+// Store represents a storage database.
+type Store interface {
+	// Put stores the key + value pair along with the (optional) tags.
+	Put(key string, value []byte, tags ...Tag) error
+
+	// Get fetches the value associated with the given key.
+	Get(key string) ([]byte, error)
+
+	// Get fetches all tags associated with the given key.
+	GetTags(k string) ([]Tag, error)
+
+	// GetBulk fetches the values associated with the given keys.
```

Comment:
Similar comment https://github.com/hyperledger/aries-framework-go/pull/2413/files#r547979499

Reply:
See https://github.com/hyperledger/aries-framework-go/pull/2426/files#diff-98db716f6d6bedd511a5f903dc6d785b6f60df130787f50612d64804a1f8195fR13 in my new PR.